src/Entity/ListBuilder/DeveloperAppListBuilder.php

```diff
@@ -320,62 +400,41 @@ public function buildRow(EntityInterface $entity) {
-      $infoRow['app_status']['data'] = $this->renderer->render($build);
-      $warningRow['info'] = [
+      $infoRow['data']['app_status']['data'] = $this->renderer->render($build);
+      $row['data']['info'] = [
         'colspan' => 3,
       ];
 
-      if ($hasRevokedCred) {
-        $args = [
-          '@developer_app' => $this->getDeveloperAppEntityDefinition()->getLowercaseLabel(),
-        ];
-        if (count($entity->getCredentials()) > 1) {
-          $warningRow['info']['data'] = $this->t(
-            'One of the credentials associated with this @developer_app is in revoked status.',
-            $args
-          );
-        }
-        else {
-          $warningRow['info']['data'] = $this->t(
-            'The credential associated with this @developer_app is in revoked status.',
-            $args
-          );
-        }
+      if ($warnings['revokedCred']) {
+        $row['data']['info']['data'] = $warnings['revokedCred'];
       }
-      elseif ($hasRevokedCredProduct || $hasPendingCredProduct) {
-        $args = [
-          '@developer_app' => $this->getDeveloperAppEntityDefinition()->getLowercaseLabel(),
-          '@apiproduct' => $this->getApiProductEntityDefinition()->getLowercaseLabel(),
-          '@status' => $hasPendingCredProduct ? $this->t('pending') : $this->t('revoked'),
-        ];
-        if (count($entity->getCredentials()) === 1) {
-          /** @var \Drupal\apigee_edge\Entity\ApiProductInterface $apiProduct */
-          $apiProduct = $this->getApiProductStorage()->load($problematicApiProductName);
-          $args['%name'] = $apiProduct->label();
-          $warningRow['info']['data'] = $this->t("%name @apiproduct associated with this @developer_app is in @status status.", $args);
-        }
-        else {
-          $warningRow['info']['data'] = $this->t("At least one @apiproduct associated with one of the credentials of this @developer_app is in @status status.", $args);
-        }
+      elseif ($warnings['revokedOrPendingCredProduct']) {
+        $row['data']['info']['data'] = $warnings['revokedOrPendingCredProduct'];
       }
     }
 
-    return $rows;
+    return $row;
   }
 
   /**
-   * Returns a rendered link to Add developer app form.
-   *
-   * @return array
-   *   Render array.
+   * {@inheritdoc}
    */
-  protected function renderAddAppLink() {
-    return Link::createFromRoute($this->t('Add @developer_app', [
-      '@developer_app' => $this->getDeveloperAppEntityDefinition()->getLowercaseLabel(),
-    ]), 'entity.developer_app.add_form', [], ['attributes' => ['class' => 'btn btn-primary btn--add-app']])->toRenderable();
+  public function buildRow(EntityInterface $entity) {
+    /** @var \Drupal\apigee_edge\Entity\DeveloperAppInterface $entity */
+    $rows = [];
+    $appNameAsCssId = $this->getUniqueCssIdForApp($entity);
+
+    // Generate info row of the current developer app.
+    $infoRowId = "{$appNameAsCssId}-info";
+    $rows[$infoRowId] = $this->buildInfoRow($entity);
```

Comment:
Why don't you just pass `$rows` as reference to both `buildInfoRow()` and `buildWarningRow()` and let them to alter the structure of rows as they want? `$infoRowId` could be calculated in for an app in a dedicated method, same for the `$warningRowId`. You would not need to pass `$rows[$infoRowId]` to the `buildWarningRow()` just the `&$rows` because you could identify the related info row by calculating the info row id for the $entity with the dedicated method.
Consequently none of these method would need to return anything, they would just directly alter the $rows if they want.